lib/polint.rb

```diff
@@ -77,39 +108,47 @@ def annotate_and_load_po
         io.write data.join
         io.flush
         io.close_write
-        data = io.read.split(/\n/)
+        data = io.read
       end
       success and return data
       die "Error while loading the PO file '#{@pofile}', aborting."
     end
 
-    def parse_data(lines)
-      current = []
-      context = []
-      next_is_fuzzy = false
-
-      lines.each do |line|
-        case line
+    def parse_data(data)
+      tree = Polint::Parser.new.parse(data)
+      tree = Polint::Transform.new.apply(tree)
 
-        when /^#, fuzzy/
-          next_is_fuzzy = true
+      die 'No Plural-Forms header found' unless tree[:headers].key?('Plural-Forms')
+      nplurals = tree[:headers]['Plural-Forms'][:nplurals]
 
-        when /^#: (.*)/
-          context << $1
+      tree[:translations].each do |translation|
+        msgid, msgid_plural = translation[:msgid][:text], translation[:msgid_plural][:text]
+        contexts = translation[:references]
+        fuzzy = translation[:flags].include?(:fuzzy)
 
-        when /^(msgid|msgid_plural)\s+"(.*)"$/
-          current << $2
-
-        when /^(msgstr(?:\[([01])\])?)\s+"(.*)"$/
-          index = $2.to_i # $2 will be nil for non-plural, so index will be 0
-          check_pair(current[index], $3, context, next_is_fuzzy)
+        if msgid_plural.nil?
+          if translation[:msgstrs].size != 1
+            error msgid, nil, contexts, "#{translation[:msgstrs].size} plurals found but none expected"
+            @errors += 1
+          end
 
+          check_pair(msgid, translation[:msgstrs][0][:text], translation[:references], fuzzy)
         else
-          current = []
-          context = []
-          next_is_fuzzy = false
+          if translation[:msgstrs].size != nplurals
+            error msgid, nil, contexts, "#{translation[:msgstrs].size} plurals found but #{nplurals} expected"
+            @errors += 1
+          end
+
+          plural_attr = (msgid_plural.scan(AttributeRe).uniq - msgid.scan(AttributeRe).uniq).first
+          translation[:msgstrs].each do |msgstr|
+            val = msgstr[:text]
+            msgid_check = val.scan(AttributeRe).include?(plural_attr) ? msgid_plural : msgid
+            check_pair(msgid_check, val, contexts, fuzzy)
+          end
```

Comment:
It's a bit convoluted, sorry :-)

the singular string doesn't have to have a numeric placeholder in, e.g.

```
msgid "One second before %{event}"
msgid_plural "%{n} seconds before %{event}"
```

So `plural_attr` is a best guess at which one is the numeric placeholder being the one that _does_ exist in `msgid_plural` but not in `msgid`. It's probably not 100% reliable but seems like a reasonable heuristic, and appears to work against all our existing translations.

Then yeah, because any translation can use the placeholders from either the singular or plural string it checks against one or the other based on whether the `msgstr` contains the placeholder. In case both `msgid` and `msgid_plural` contain the same placeholders then it'll check against `msgid` but it shouldn't really matter.